display_information/update-graphics/src/main/java/com/esri/samples/update_graphics/UpdateGraphicsSample.java

```diff
@@ -167,39 +156,52 @@ public void start(Stage stage) {
 
       mapView.setOnMouseClicked(e -> {
         if (e.getButton() == MouseButton.PRIMARY && e.isStillSincePress()) {
+          // set the cursor to default
+          mapView.setCursor(Cursor.DEFAULT);
+
           // clear any selected graphic
           graphicsOverlay.clearSelection();
 
-          // create a point from location clicked
-          Point2D mapViewPoint = new Point2D(e.getX(), e.getY());
-
-          if (isUpdateLocationActive) {
-            // add new location to selected graphic
-            Point mapPoint = mapView.screenToLocation(mapViewPoint);
-            selectedGraphic.setGeometry(mapPoint);
-            isUpdateLocationActive = false;
-          } else {
-            // identify the graphic that was selected
-            identifyGraphics = mapView.identifyGraphicsOverlayAsync(graphicsOverlay, mapViewPoint, 10, false);
-
-            identifyGraphics.addDoneListener(() -> {
-              try {
-                if (!identifyGraphics.get().getGraphics().isEmpty()) {
-                  // store the selected graphic
-                  selectedGraphic = identifyGraphics.get().getGraphics().get(0);
-                  selectedGraphic.setSelected(true);
-                  String style = ((SimpleMarkerSymbol) selectedGraphic.getSymbol()).getStyle().toString();
-                  symbolBox.getSelectionModel().select(style);
-
-                  enableUI(false);
-                } else {
-                  enableUI(true);
-                }
-              } catch (Exception x) {
-                new Alert(Alert.AlertType.ERROR, "Error identifying clicked graphic").show();
+          // create a point where the user clicked
+          mapViewPoint = new Point2D(e.getX(), e.getY());
+
+          // identify graphics on the graphics overlay
+          ListenableFuture<IdentifyGraphicsOverlayResult> identifyGraphics =
+            mapView.identifyGraphicsOverlayAsync(graphicsOverlay, mapViewPoint, 10, false);
+
+          identifyGraphics.addDoneListener(() -> {
+            try {
+              if (!identifyGraphics.get().getGraphics().isEmpty()) {
+                // get the first identified graphic
+                identifiedGraphic = identifyGraphics.get().getGraphics().get(0);
+                // select the identified graphic
+                identifiedGraphic.setSelected(true);
+                // update the drop down box with the identified graphic's current symbol
+                String style = ((SimpleMarkerSymbol) identifiedGraphic.getSymbol()).getStyle().toString();
+                symbolBox.getSelectionModel().select(style);
+                // show the UI
+                disableUI(false);
```

Comment:
A few things in this suggestion. 

1. When the sample loads, and the user clicks and drags on the map view, a null pointer exception is thrown. This continues until a user clicks to select a graphic, and then it's not seen again. We could handle that by showing an error in the mapView.setOnMouseDragged listener, or, by simply moving that listener in the identifyGraphics.addDoneListener (see suggestion).

The benefit to this approach is that we know the identifiedGraphic should never be null (since it's already identified), and a null pointer shouldn't be able to be thrown. Also when the sample loads, a user could click and drag the map view if they wanted to before it gets locked when selecting a graphic.

2.  If a user selects a graphic but clicks and drags on the other side of the screen, the point is dragged. I'd expect the graphic only to be dragged if the cursor was actually on the graphic. I think Jon has suggested something similar above to address this but might have too many code arms and legs to follow phrase! So to make it clear to the user that the graphic is still selected and in drag mode, I've suggested a simple `Cursor.CLOSED_HAND` here to make it visually obvious we're still dragging. 

```suggestion
                disableUI(false);

                // enable dragging of the identified graphic to move its location
                mapView.setOnMouseDragged(event -> {

                    if (identifiedGraphic.isSelected() && identifiedGraphic != null) {
                      // set the cursor to closed hand to indicate graphic dragging is active
                      mapView.setCursor(Cursor.CLOSED_HAND);
                      // create a point from the dragged location
                      mapViewPoint = new Point2D(event.getX(), event.getY());
                      Point mapPoint = mapView.screenToLocation(mapViewPoint);
                      // update the location of the graphic to the dragged location
                      identifiedGraphic.setGeometry(mapPoint);
                    } else new Alert(Alert.AlertType.ERROR, "Error selecting graphic").show();
                });
```

Reply:
Alternatively, it is possible to simultaneously click and drag the graphic with the following. 

However, I had to bump the tolerance up quite high since there is a bit of lag with the drag and if the mouse travels faster than the graphic, it leaves it behind. With this bigger tolerance though, more than one graphic can be selected and it results in a fun dance between the two graphics moving...if you'd like to take this code and work around it to avoid that behaviour, see what you think: it might take a fair bit of time, I've had to leave it for now. Otherwise the workflow I've suggested above is just fine! 

```java
      mapView.setOnMouseDragged(e -> {
        if (e.getButton() == MouseButton.PRIMARY) {
//           create a point where the user clicked
          mapViewPoint = new Point2D(e.getX(), e.getY());

          // identify graphics on the graphics overlay
          ListenableFuture<IdentifyGraphicsOverlayResult> identifyGraphics =
            mapView.identifyGraphicsOverlayAsync(graphicsOverlay, mapViewPoint, 50, false);

          identifyGraphics.addDoneListener(() -> {
            try {
              if (!identifyGraphics.get().getGraphics().isEmpty()) {
                // get the first identified graphic
                identifiedGraphic = identifyGraphics.get().getGraphics().get(0);
                // select the identified graphic
                identifiedGraphic.setSelected(true);
                // update the drop down box with the identified graphic's current symbol
                String style = ((SimpleMarkerSymbol) identifiedGraphic.getSymbol()).getStyle().toString();
                symbolBox.getSelectionModel().select(style);
                // show the UI
                disableUI(false);

                // enable dragging of the identified graphic to move its location
                if (identifiedGraphic.isSelected() && identifiedGraphic != null) {
                  // set the cursor to closed hand to indicate graphic dragging is active
//                      mapView.setCursor(Cursor.CLOSED_HAND);
                  // create a point from the dragged location
                  mapViewPoint = new Point2D(e.getX(), e.getY());
                  Point mapPoint = mapView.screenToLocation(mapViewPoint);
                  // update the location of the graphic to the dragged location
                  identifiedGraphic.setGeometry(mapPoint);
                } else new Alert(Alert.AlertType.ERROR, "Error selecting graphic").show();

                // clear any selected graphic when clicking on the map view
                mapView.setOnMouseClicked(mouseEvent -> {
                  if (mouseEvent.isStillSincePress()) {
                    graphicsOverlay.clearSelection();
                  }
                });

              } else {
                disableUI(true);
              }
            } catch (Exception x) {
              new Alert(Alert.AlertType.ERROR, "Error identifying clicked graphic").show();
            }
          });
        }
      });
```